shapeshift/lib
# feat(chain-adapters:UtxoBaseAdapter): allow UTXOs selection from `from` param

#### Description

- [x] Partial sends
- [ ] ~~Send max - do we want to handle this for consistency? Given our use-case for savers, we shouldn't need this~~

This PR brings a new `from` input param in `buildSendTransaction` in `getFeeData` and `buildSendTransaction`, being passed down as a new input param in `utxoSelect` and allowing us to filter UTXOs to only the ones for the passed address.

This effectively allows us to ensure the right UTXO inputs are used when building Txs for Thorchain savers.

Tested by monkey-patching a `from` param in `buildSendTransaction`, effectively mocking sending to the address / pubkey (*NOT* an xpub) that was previously used as an output to deposit into savers.

The current philosophy of this added logic is to "let it crash" in case the coinselect algorithm fails to create an output and/or the broadcast fails, so that web can handle the error, do a regular send (without a `from` param) to the original deposit address to create a valid UTXO, and build/broadcast a Tx again.

#### Testing

- View your UTXOs in `https://blockchair.com/litecoin/outputs?q=recipient(<address>),is_spent(false)#` or by setting a breakpoint in https://github.com/shapeshift/lib/blob/ce34273b11c481665b6f28bd5db0171be3eda210/packages/chain-adapters/src/utxo/utxoSelect/index.ts#L23
- In web, call `buildSendTransaction` with a `from` address that will produce valid outputs
- Ensure the Tx is broadcasted

#### Screenshots

<img width="850" alt="image" src="https://user-images.githubusercontent.com/17035424/210660722-0e172e76-83ab-4b17-8dad-41bc7cb9bad8.png">
